File: app/src/main/java/com/spartons/androiddownloadmanager/MainActivity.java
package com.spartons.androiddownloadmanager;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.example.downloadmanagers_demo.R;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private static final String IMAGE_DOWNLOAD_PATH = "http://globalmedicalco.com/photos/globalmedicalco/9/41427.jpg";
    private static final String SONG_DOWNLOAD_PATH = "https://cloudup.com/files/inYVmLryD4p/download";
    private static final int WRITE_EXTERNAL_STORAGE_REQUEST_CODE = 54654;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        findViewById(R.id.downloadImageButton).setOnClickListener(this);
        findViewById(R.id.downloadSongButton).setOnClickListener(this);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, WRITE_EXTERNAL_STORAGE_REQUEST_CODE);
            return;
        }
        com.spartons.androiddownloadmanager.DirectoryHelper.createDirectory(this);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.downloadImageButton: {
                startService(com.spartons.androiddownloadmanager.DownloadSongService.getDownloadService(this, IMAGE_DOWNLOAD_PATH, com.spartons.androiddownloadmanager.DirectoryHelper.ROOT_DIRECTORY_NAME.concat("/")));
                break;
            }
            case R.id.downloadSongButton: {
                startService(com.spartons.androiddownloadmanager.DownloadSongService.getDownloadService(this, SONG_DOWNLOAD_PATH, com.spartons.androiddownloadmanager.DirectoryHelper.ROOT_DIRECTORY_NAME.concat("/")));
                break;
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == WRITE_EXTERNAL_STORAGE_REQUEST_CODE) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED)
                com.spartons.androiddownloadmanager.DirectoryHelper.createDirectory(this);
        }
    }
}

File: app/src/main/java/com/spartons/androiddownloadmanager/MainActivity.kt
package com.spartons.androiddownloadmanager

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import com.example.downloadmanagers_demo.R
import com.spartons.androiddownloadmanager.DirectoryHelper.Companion.ROOT_DIRECTORY_NAME
import com.spartons.androiddownloadmanager.DownloadSongService.getDownloadService
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private lateinit var permissionsDelegate: PermissionsDelegate
    private lateinit var networkDownloadsGateway: DownloadsGateway

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        permissionsDelegate = PermissionsDelegate(DirectoryHelper(this), ActivityWrapper(this))
        networkDownloadsGateway = NetworkDownloadsGateway(this)

        setListeners()
    }

    private fun setListeners() {
        downloadImageButton.setOnClickListener {
            requestDownloadImage()
        }
        downloadSongButton.setOnClickListener {
            requestDownloadSong()
        }
    }

    private fun requestDownloadImage() {
        permissionsDelegate.requestPermission(object : PermissionsDelegate.Listener {
            override fun onPermissionGranted() {
                doDownloadImage()
            }
        })
    }

    private fun doDownloadImage() {
        startService(getDownloadService(this, PDF_DOWNLOAD_PATH, "$ROOT_DIRECTORY_NAME/"))
    }

    private fun requestDownloadSong() {
        permissionsDelegate.requestPermission(object : PermissionsDelegate.Listener {
            override fun onPermissionGranted() {
                doDownloadSong()
            }
        })
    }

    private fun doDownloadSong() {
        startService(getDownloadService(this, SONG_DOWNLOAD_PATH, "$ROOT_DIRECTORY_NAME/"))
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        permissionsDelegate.onRequestPermissionsResult(requestCode, grantResults)
    }


    companion object {
        private const val PDF_DOWNLOAD_PATH =
            "https://www.goldmansachs.com/insights/pages/gs-research/taking-the-heat/report.pdf"
        private const val SONG_DOWNLOAD_PATH = "https://cloudup.com/files/inYVmLryD4p/download"
    }
}
